algorithms/linfa-ensemble/src/gradient_boost/algorithm.rs

```diff
@@ -0,0 +1,323 @@
+        // final_predictions /= self.num_trees as f64;
+        final_predictions
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use approx::assert_relative_eq;
+    use ndarray::{array, Array1, Array2}; // For floating-point assertions
+
+    // #[test]
+    // fn test_gradientboost_regressor() {
+    //     let features = Array2::from_shape_vec((4, 1), vec![1.0, 2.0, 4.0, 5.0]).unwrap();
+    //     let targets = Array1::from_vec(vec![1.1, 1.9, 3.9, 5.1]);
+    //     let mut regressor = DecisionTreeRegressor::new(3, 1);
+
+    //     if features.nrows() != targets.len() {
+    //         panic!("Feature and target count mismatch");
+    //     }
+
+    //     regressor.fit(&features, &targets);
+    //     let predictions = regressor.predict(&features);
+    //     let rmse = calculate_rmse(&targets, &predictions);
+
+    //     println!("RMSE: {:?}", rmse);
+    //     assert_relative_eq!(rmse, 0.0, epsilon = 0.1);
+    // }
+
+    // fn calculate_rmse(actual: &Array1<f64>, predicted: &Array1<f64>) -> f64 {
+    //     let errors = actual - predicted;
+    //     let mse = errors.mapv(|e| e.powi(2)).mean().unwrap();
+    //     mse.sqrt()
+    // }
+
+    fn load_iris_data() -> (Array2<f64>, Array1<f64>) {
+        // Load the dataset
+        let dataset = linfa_datasets::iris();
+
+        // Extract features; assuming all rows and all but the last column if last is target
+        let features = dataset.records().clone();
+
+        // Assuming the target are the labels, we need to convert them or use them as is depending on the use case.
+        // If you need to predict a feature, then split differently as shown in previous messages.
+        // Here we just clone the labels for the demonstration.
+        let targets = dataset.targets().mapv(|x| x as f64);
+
+        (features, targets)
+    }
+
+    #[test]
+    fn test_gradient_boost_with_iris() {
+        let (features, targets) = load_iris_data();
+
+        let mut gb_regressor = GBRegressor::new(50, 0.1, 4, 10, 3);
+        gb_regressor.fit(&features, &targets);
+        let predictions = gb_regressor.predict(&features);
+
+        // Define a tolerance level
+        let tolerance = 0.01; // Tolerance level for correct classification
+        let mut correct = 0;
+        let mut incorrect = 0;
+
+        // Count correct and incorrect predictions
+        for (&actual, &predicted) in targets.iter().zip(predictions.iter()) {
+            if (predicted - actual).abs() < tolerance {
+                correct += 1;
+            } else {
+                incorrect += 1;
+            }
+        }
+
+        println!("Correct predictions: {}", correct);
+        println!("Incorrect predictions: {}", incorrect);
+
+        let rmse = (&predictions - &targets)
+            .mapv(|a| a.powi(2))
+            .mean()
+            .unwrap()
+            .sqrt();
+
+        println!("RMSE: {:?}", rmse);
+    }
+}
+
+
+// fn softmax(residuals: &mut Vec<f64>) -> Result<()> {
```

Comment:
remove the comments if not needed

Reply:
Done, it was just the code which I had written in the beginning.